### PR TITLE
Carets - Canaan

### PR DESCRIPTION
**Pull request by canaanwest**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   When I began Wave 3 and had to start thinking about reservation blocks, I wanted to use inheritance to say that a Block would be a kind of Reservation. However, because of the way I'd structured my reservation methods (to handle a single argument of class Room), almost all of the methods I'd written for dealing with reservations were useless for my blocks. For this reason, I decided to collapse Reservations and Blocks into a single category. In this way, I was able to simplify my code by having methods that could deal with dynamic reservation types.   |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  Although it's definitely not perfect, as the week progressed, I started to see how separations of concerns / single responsibility makes code easier to modify. I began to break things up more towards the end of the week as I began reading about OOP design techniques, and I saw how having a new class for something like "Nights" (which, in my program, is the class that handles dates of check-in, check-out, etc.), which was necessary for talking about reservations but also pervasive throughout my program's methods, really simplified/clarified many of my methods throughout the program.  |
| Describe a _nominal test_ that you wrote for this assignment. |  it "Can create a reservation with more than one room" was an example of a nominal test to make sure the reservation class could handle reservations for multiple rooms  |
| Describe an _edge case test_ that you wrote for this assignment. |  In contrast, "it "will not double-book a room in a block"" was an edge case that tested that the program raised an argument error if the same room was added to a block more than once.  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did ok at this. I definitely pseudocoded an outline as soon as I read the instructions for the tests in each wave. I probably could have done a little better with the code itself  |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  end
+
+
+  def make_reservation(check_in, check_out, rooms)
+    unavailable_rooms = []
+
+    if room_flagged_available?(check_in, check_out, rooms, unavailable_rooms)
+      @reservations.push(Reservation.new(check_in, check_out, rooms))
+    else
+      begin
```
Comment (tildeee) [inline lib/hotel.rb]: Refer back to our [exception handling lecture](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/exception-handling.md) about how to set up this exception handling

File: specs/nights_spec.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+require_relative '../lib/nights'
+require_relative 'spec_helper'
+
+describe "NIGHTS" do
+  before do
+    @summer_somewhere_nights = Nights.new("January 4, 2018", "January 11, 2018")
+  end
+
+  describe "#INITIALIZE" do
```
Comment (tildeee) [inline specs/nights_spec.rb:9]: It was a good call to make a `Nights` class which would deal with the logic of thinking about a date range. I would've liked to see more test coverage around the logic that belongs to this (aka valid dates)

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | x
**Wave 1** |
List rooms | x
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | x
Test coverage | x
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | x
Test coverage | x
**Wave 3** |
Create a block of rooms | ?
Check if a block has rooms | ?
Reserve a room from a block | ?
Test coverage | x
**Additional Feedback** |

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      end
+    end
+  end
+
+  def room_flagged_available?(check_in, check_out, rooms, unavailable_rooms)
+    flag = true
+    rooms.each do |room|
+      if !(rooms_available(check_in, check_out).include?(room))
+        unavailable_rooms.push(room.number)
+        flag = false
```
Comment (tildeee) [inline lib/hotel.rb:127]: could probably refactor this method quite a bit

**Comment by tildeee** [comment]
Overall good hotel, I think it was a good design decision to pull out the `Nights` class to deal with date logic. This helped keep your `Hotel` and `Reservation`s become more independent of each other.

I think your interpretation of the requirements for creating a Block of rooms and then later reserving them got a little confused in your implementation; from what I can see, you implemented it so that you can reserve multiple rooms at a given time.
